[RFC] Transition frontier listener

Schmavery opened the pull request.

bkase commented on rfcs/0000-frontier-listener.md:
```diff
@@ -0,0 +1,76 @@
+  type t
+  val add_breadcrumb : t -> Breadcrumb.t -> unit;
+  val remove_breadcrumb : t -> Breadcrumb.t -> unit;
+end
```
I find hiding information in closures something that makes code harder to read and trace -- so I prefer the first way.

Schmavery replied:
I've eventually managed to prototype something out with first-class modules but I would be fairly on board with switching to closures as it seems like it would simplify the gymnastics involved in implementing/reading this code..

nholland94 commented on rfcs/0000-transition-frontier-listener.md:
````diff
@@ -0,0 +1,84 @@
+
+The transition frontier would then expose a function for registering listeners. We probably have to do some sort of GADT/ first-class module magic to make this work (???)
+```
+val add_frontier_listener : (module Transition_frontier_listener_intf with type t = 'a) -> 'a -> unit
````
I'm not so sure this should be a dynamic action. While we want to be able to easily add new listeners to the transition frontier over time, I don't see a use case for dynamic selection of which listeners to attach. I think we should actually hard code the set of listeners will be attached as a constant, and make a wrapper layer which contains both the transition frontier and all the listeners under unique names in a record. This way, it's easy to consume the listeners and still easy to extend the code to add a new one.

nholland94 commented:
Oh, one more thing. We sort of discussed this, but what do you think about formalizing this using the same diff abstraction we need for the persistence layer? For example, imagine we have a diff type `'a diff` which has the variants `Add of 'a | Remove of 'a | Set_root of 'a`. The interface would then be a function which takes a diff with the type `Breadcrumb.t diff`. In the case of the persistent listener, it would map that diff into a `Minimal_breadcrumb.t diff` or something. It's not really necessary for the other listeners, but I think it makes a nice reusable abstraction. What do you think though?

Schmavery commented:
Trying my best to incorporate feedback. The lifetime/ownership of these objects is the part that still the most unclear to me, but here's another design that might better match what you are imagining.
This goes with the assumption that there won't be very many of these views and that they'd be completely static, but lmk if it's way off.

Here are the signatures we could add to Protocols.Coda_transition_frontier:
```ocaml
module type Transition_frontier_iview = sig
  type t
  val create: unit -> t
  val handle_diff: t -> node frontier_diff -> unit
end

type 'a frontier_diff =
  | Add of 'a
  | Remove of 'a
  | Set_root of 'a
  | Destroy

type iviews =
  { snark_pool_refcount: Snark_pool_refcount.t
  ; transaction_pool_refcount: Transaction_pool_refcount.t
  ; frontier_persister: Frontier_persister.t}
```

This would be combined with the addition of functions like this to `Transition_frontier.ml`:
```ocaml
module Transition_frontier = struct
...
  let notify_iviews_of_diff iviews diff =
    Snark_pool_refcount.handle_diff iviews.snark_pool_refcount diff;
    Transaction_pool_refcount.handle_diff iviews.transaction_pool_refcount diff;     
    Frontier_persister.handle_diff iviews.frontier_persister diff;

  (* Called inside Transition_frontier.create *)
  let create_iviews () =
    { snark_pool_refcount = Snark_pool_refcount.create (),
    ; transaction_pool_refcount = Transaction_pool_refcount.create (),
    ; frontier_persister = Frontier_persister.create ()}
end
```

Here's a rough example of what an implementation of `Transition_frontier_iview` could look like for the Snark_pool reference counter.
```ocaml
  module Snark_pool_refcount = struct
    module type S = Transition_frontier_iview
    type t = {ref_table: Work.t int Hashtbl.t
              ; removed: Work.t Hash_set.t }

    let create () =
      { ref_table = Hashtbl.create ()
      ; removed = Hash_set.create ()}

    let handle_diff t diff = match diff with
      | Set_root node (* ??? *)
      | Add node ->
        let work = somehow_get_work node in
        List.iter(work, ~f:(fun w -> Hashtbl.update t w ~f:(function
          | Some n -> n + 1
          | None -> 1)))
      | Remove node ->
        let work = somehow_get_work node in
        List.iter(work, ~f:(fun w -> Hashtbl.change t w ~f:(function
          | None | Some 1 -> Hash_set.add t.removed w; None
          | Some n -> Some (n - 1))))
      | Destroy -> ()
  end
```

bkase commented:
Your changes look good to me, here are my three small things to add:

1. I dislike the `iview` name (what is i? acronyms!) `extension` is probably what I'd vote for.
2. If you put `[@@deriving fields]` on the record with the extensions, we can invoke the listeners with a `fold` over the fields and get compile errors if we forget to update when we add a new extension (see `lib/daemon_rpcs/daemon_rpcs.ml` for an example of this)
3. Another nit, prefer nested modules with `type t` to named types (see `docs/style_guide.md`)

cmr commented:
The extensions should receive as an argument the transition frontier that was subject to the event?

When the extension methods are called, is the corresponding action already fully completed?

Must we block the relevant transition frontier methods on calling into the extensions, or can they be scheduled to occur later?

cmr commented:
> Must we block the relevant transition frontier methods on calling into the extensions, or can they be scheduled to occur later?

If they must block, we should lift the extension methods into `Deferred` so that we can do asynchronous work without having a `don't_wait_for` that potentially invalidates any code that depending on blocking-until-extensions-are-done.

nholland94 commented:
> The lifetime/ownership of these objects is the part that still the most unclear to me, but here's another design that might better match what you are imagining.

The transition frontier controller "owns" both the transition frontier and the extensions to it. The lifetime of these two values should be tied to each other, which is to say that, when the transition frontier is created, the extensions, and when the transition frontier is destroyed, the extensions should be as well (destruction is implicit though).

nholland94 commented:
Looks like this is converging in the correct direction. I would second what @bkase said about using `[@@deriving fields]` as a way of constraining the application of diffs to extensions.

To respond to @cmr:

> The extensions should receive as an argument the transition frontier that was subject to the event?

Yes

> When the extension methods are called, is the corresponding action already fully completed?

Yes

> Must we block the relevant transition frontier methods on calling into the extensions, or can they be scheduled to occur later?

Diffs are generated by the relevant transition frontier methods, so yes. They could be split later if need be, but I don't see any reason to require that now. The processor thread can perform the transition frontier actions and defer another thread to handle passing the diff to the extensions if we need to split it up, but I think that the extensions we need will be fast enough to update in thread that we can just do that.

Schmavery commented:
@cmr @nholland94 Thanks for feedback, one question about this:

> > The extensions should receive as an argument the transition frontier that was subject to the event?
> 
> Yes

The original idea was to just pass the diff (ie, the breadcrumb that was removed/added). Is that not enough information and we think that most extensions will need to able to also access the transition frontier itself, or is there another reason?

enolan commented:
We added the extensions mechanism and never merged this rfc... and it's going to be different again when my changes are merged
